### ISSUE TITLE
Feature Request: Add support for Phi-3.5 MoE and Vision Instruct

### ISSUE THREAD

**Issue by YorkieDev** [ggerganov/llama.cpp#9119]
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

Microsoft has recently dropped two new models in the Phi Family. 

3.5 MoE: https://huggingface.co/microsoft/Phi-3.5-MoE-instruct
3.5 Vision: https://huggingface.co/microsoft/Phi-3.5-vision-instruct

It would be nice to see support added to llama.cpp for these two models. 

### Motivation

Supporting all model releases so the wider community can enjoy these great free models.

**Comment by curvedinf** [comment]
MoE looks promising. Any word on how complex it is to add support for?

**Comment by JackCloudman** [comment]
Is someone working on it? :pray:

**Comment by simsi-andy** [comment]
Especially vision would be worth it. But I lack the knowledge to do smth. like this.

**Comment by mounta11n** [comment]
Yes, the vision model is surprisingly good. As a gguf format under llama.cpp, this would open up undreamt-of possibilities

![Bildschirmfoto_20240821_213502](https://github.com/user-attachments/assets/34b17c53-2b8b-4339-8292-fc974c2972d7)

**Comment by foldl** [comment]
[ChatLLM.cpp](https://github.com/foldl/chatllm.cpp) supports Phi-3.5 MoE model now.

For developers: MoE Sparse MLP is the same as the one used in Mixtral.

**Comment by ayttop** [comment]
https://github.com/foldl/chatllm.cpp

| [Supported Models](https://github.com/foldl/chatllm.cpp/blob/master/docs/models.md) | [Download Quantized Models](https://github.com/foldl/chatllm.cpp/blob/master/docs/quick_start.md#download-quantized-models) |

What's New:

2024-08-28: Phi-3.5 Mini & MoE


Inference of a bunch of models from less than 1B to more than 300B, for real-time chatting with [RAG](https://github.com/foldl/chatllm.cpp/blob/master/docs/rag.md) on your computer (CPU), pure C++ implementation based on [@ggerganov](https://github.com/ggerganov)'s [ggml](https://github.com/ggerganov/ggml).

| [Supported Models](https://github.com/foldl/chatllm.cpp/blob/master/docs/models.md) | [Download Quantized Models](https://github.com/foldl/chatllm.cpp/blob/master/docs/quick_start.md#download-quantized-models) |

What's New:

2024-08-28: Phi-3.5 Mini & MoE

**Comment by ayttop** [comment]
https://huggingface.co/microsoft/Phi-3.5-MoE-instruct/discussions/4

microsoft/Phi-3.5-MoE-instruct
convert to gguf
gguf

**Comment by Dampfinchen** [comment]
Pretty sad to see no support for Phi 3.5 MoE in llama.cpp. Sure, it might have dry writing and is very censored, but in assistant tasks it's much better than all the smaller models combined. It truly has 70B quality in just 6.6B active parameters so its much easier to run than even G2 27B (which it beats according to benchmarks).

**Comment by sourceholder** [comment]
@Dampfinchen, have you found _any_ way to run Phi 3.5 MoE locally? I'm open to try out alternatives to llama.cpp.

**Comment by arnesund** [comment]
Also eager to get Phi 3.5-Vision support. Most accurate photo and screenshot descriptions I've seen so far.

**Comment by EricLBuehler** [comment]
@Dampfinchen @sourceholder @arnesund if you are interested in running Phi 3.5 MoE or Phi 3.5 vision with alternatives to llama.cpp, perhaps you could check out mistral.rs.

Just a quick description:

We have support for Phi 3.5 MoE (docs & example: https://github.com/EricLBuehler/mistral.rs/blob/master/docs/PHI3.5MOE.md) and Phi 3.5 vision (docs & examples: https://github.com/EricLBuehler/mistral.rs/blob/master/docs/PHI3V.md).

All models can be run with CUDA, Metal, or CPU SIMD acceleration. We have Flash Attention and PagedAttention support for increased inference performance, and support in-situ quantization in GGUF and HQQ formats.

If you are using the OpenAI API, you can use the provided OpenAI-compatible (superset, we have things like min-p, DRY, etc) HTTP server. There is also a Python package. For Phi 3.5 MoE and other text models, there is also an interactive chat mode.

**Comment by Dampfinchen** [comment]
Thank you, but I and many others rather wait for official support.

I wonder what's the holdup? Shouldn't it be possible to copy a lot of the code from Mixtral to Phi 3.5 MoE given they have a pretty similar architecture with two experts?

**Comment by Thellton** [comment]
> Thank you, but I and many others rather wait for official support.
> 
> I wonder what's the holdup? Shouldn't it be possible to copy a lot of the code from Mixtral to Phi 3.5 MoE given they have a pretty similar architecture with two experts?

no one's taken the task up yet sadly. there's presently work being done on Phi-3.5 Vision Instruct though which is something to look forward to considering the reported vision understanding that the model has.

**Comment by ayttop** [comment]
phi-3.5-moe-instruct gguf lamacpp???????????????????????

**Comment by vaibhav1618** [comment]
Strange why no one is looking into this. MoE seems to be the best model currently that can run on consumer-grade cpu

**Comment by sourceholder** [comment]
@vaibhav1618, FYI - Deepseek V2 Lite (16B) is another good MoE model.   2.4B activated params.

**Comment by ThiloteE** [comment]
Phi-3.5 MoE seems to be based on https://huggingface.co/microsoft/GRIN-MoE/tree/main. Maybe their technical report at https://arxiv.org/abs/2409.12136 can help at identifying differences to other MoE architectures, which should ease adoption in llama.cpp.

**Comment by yueshen-intel** [comment]
> there's presently work being done on Phi-3.5 Vision Instruct though which is something to look forward to considering the reported vision understanding that the model has

I'm wondering where's the work being done on Phi-3.5 Vision Instruct ? Much thanks!

**Comment by limingchina** [comment]
> @Dampfinchen @sourceholder @arnesund if you are interested in running Phi 3.5 MoE or Phi 3.5 vision with alternatives to llama.cpp, perhaps you could check out mistral.rs.
> 
> Just a quick description:
> 
> We have support for Phi 3.5 MoE (docs & example: https://github.com/EricLBuehler/mistral.rs/blob/master/docs/PHI3.5MOE.md) and Phi 3.5 vision (docs & examples: https://github.com/EricLBuehler/mistral.rs/blob/master/docs/PHI3V.md).
> 
> All models can be run with CUDA, Metal, or CPU SIMD acceleration. We have Flash Attention and PagedAttention support for increased inference performance, and support in-situ quantization in GGUF and HQQ formats.
> 
> If you are using the OpenAI API, you can use the provided OpenAI-compatible (superset, we have things like min-p, DRY, etc) HTTP server. There is also a Python package. For Phi 3.5 MoE and other text models, there is also an interactive chat mode.

@EricLBuehler , do you recommend some frontend app to use mistral.rs?

**Comment by sinand99** [comment]
+1 for MoE support.

**Comment by skylake5200** [comment]
> Also eager to get Phi 3.5-Vision support. Most accurate photo and screenshot descriptions I've seen so far.

+1 for Phi 3.5-Vision support.

**Comment by khemchand-zetta** [comment]
Please  checkout the PR https://github.com/ggml-org/llama.cpp/pull/17687 for phi-3.5 Vision Instruct support
